src/nodes.js

```diff
@@ -143,10 +154,44 @@ var registerChild = function(parent, key, child) {
 };
 
 
+/**
+ * Marks the key map for a given Element as invalid. It will be cleaned the
+ * next time it is needed.
+ * @param {!Element} el
+ */
+var invalidateKeyMap = function(el) {
+  getKeyMap(el).isValid = false;
+};
+
+
+/**
+ * Cleans a key map for a given Element, removing any nodes that are no longer
+ * in the DOM.
+ * @param {!Element} el
+ */
+var cleanKeyMap = function(el) {
+  var keyMap = getData(el).keyMap;
+
+  if (!keyMap || keyMap.isValid) {
+    return;
+  }
+
+  for (var key in keyMap) {
+    if (!keyMap[key].parentNode) {
+      delete keyMap[key];
```

Comment:
I don't see any performance difference in my own test re deleting vs. book-keeping. It doesn't seem to make sense that it would make a difference either. De-opt when deleting only applies to objects that are treated as typed objects and not as generic maps. It could be in your benchmark they are being treated as objects with offsets because you are using very uniform keys.